app/server-test/src/test/scala/org/bitcoins/server/WebsocketTests.scala

```diff
@@ -0,0 +1,274 @@
+package org.bitcoins.server
+
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.model.ws.{
+  Message,
+  TextMessage,
+  WebSocketRequest,
+  WebSocketUpgradeResponse
+}
+import akka.stream.scaladsl.{Flow, Keep, Sink, Source}
+import org.bitcoins.cli.{CliCommand, Config, ConsoleCli}
+import org.bitcoins.commons.jsonmodels.ws.{WalletNotification, WalletWsType}
+import org.bitcoins.commons.jsonmodels.ws.WalletNotification.{
+  BlockProcessedNotification,
+  NewAddressNotification,
+  TxBroadcastNotification,
+  TxProcessedNotification
+}
+import org.bitcoins.commons.serializers.{Picklers, WsPicklers}
+import org.bitcoins.core.currency.Bitcoins
+import org.bitcoins.core.protocol.BitcoinAddress
+import org.bitcoins.core.protocol.transaction.Transaction
+import org.bitcoins.core.wallet.fee.SatoshisPerVirtualByte
+import org.bitcoins.crypto.DoubleSha256DigestBE
+import org.bitcoins.testkit.server.{
+  BitcoinSServerMainBitcoindFixture,
+  ServerWithBitcoind
+}
+import org.bitcoins.testkit.util.AkkaUtil
+
+import scala.concurrent.duration.DurationInt
+import scala.concurrent.{Future, Promise}
+
+class WebsocketTests extends BitcoinSServerMainBitcoindFixture {
+
+  behavior of "Websocket Tests"
+
+  val endSink: Sink[WalletNotification[_], Future[Seq[WalletNotification[_]]]] =
+    Sink.seq[WalletNotification[_]]
+
+  val sink: Sink[Message, Future[Seq[WalletNotification[_]]]] = Flow[Message]
+    .map {
+      case message: TextMessage.Strict =>
+        //we should be able to parse the address message
+        val text = message.text
+        val notification: WalletNotification[_] =
+          upickle.default.read[WalletNotification[_]](text)(
+            WsPicklers.walletNotificationPickler)
+        notification
+      case msg =>
+        fail(s"Unexpected msg type received in the sink, msg=$msg")
+    }
+    .toMat(endSink)(Keep.right)
+
+  def buildReq(conf: BitcoinSAppConfig): WebSocketRequest = {
+    WebSocketRequest(s"ws://localhost:${conf.wsPort}/events")
+  }
+
+  it must "receive updates when an address is generated" in {
+    serverWithBitcoind =>
+      val ServerWithBitcoind(_, server) = serverWithBitcoind
+      val cliConfig = Config(rpcPortOpt = Some(server.conf.rpcPort))
+      //start the websocket
+      val f: Flow[
+        Message,
+        Message,
+        (Future[Seq[WalletNotification[_]]], Promise[Option[Message]])] = {
+        Flow
+          .fromSinkAndSourceCoupledMat(sink, Source.maybe[Message])(Keep.both)
```

Comment:
might be nicer to use `Source.never` if I understand what's going on here correctly: https://doc.akka.io/docs/akka/current/stream/operators/Source/never.html

Reply:
Unless you're explicitly using `Source.maybe` in place of a kill-switch, in which case I guess I'd say using `Source.never` with an explicit killswitch flow would be more idiomatic but not really a big deal

Reply:
ditto below